### PR TITLE
streamingccl/logical: remove consumer buffer

### PR DESCRIPTION
**Pull request by dt**
Release note: none.
Epic: none.

### CODE REVIEW

File: pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go [hunk 1]
```diff
@@ -686,24 +503,24 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
...
 	lrw.metrics.FlushHistNanos.RecordValue(flushTime)
 	lrw.metrics.FlushRowCountHist.RecordValue(keyCount)
 	lrw.metrics.FlushBytesHist.RecordValue(byteCount)
 	lrw.metrics.IngestedLogicalBytes.Inc(byteCount)
-	lrw.metrics.CommitLatency.RecordValue(timeutil.Since(b.buffer.minTimestamp.GoTime()).Nanoseconds())
-	lrw.metrics.IngestedEvents.Inc(int64(len(b.buffer.curKVBatch)))
 
-	releaseBuffer(b.buffer)
+	firstKeyTS := kvs[0].Value.Timestamp.GoTime()
+	lrw.metrics.CommitLatency.RecordValue(timeutil.Since(firstKeyTS).Nanoseconds())
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:520]: If we care a lot about this difference, we could have the batch return the min timestamp via batchStats and then record it on a per-batch (or chunk) basis.

File: pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go [hunk 2]
```diff
@@ -275,19 +237,12 @@ func (lrw *logicalReplicationWriterProcessor) Start(ctx context.Context) {
...
 		}
 		return nil
 	})
-	lrw.workerGroup.GoCtx(func(ctx context.Context) error {
-		defer close(lrw.checkpointCh)
-		if err := lrw.flushLoop(ctx); err != nil {
-			lrw.sendError(errors.Wrap(err, "flush loop"))
-		}
-		return nil
-	})
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:290]: There is a block comment on this function we might want to update.

File: pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go [hunk 3]
```diff
@@ -479,35 +389,10 @@ func (lrw *logicalReplicationWriterProcessor) handleEvent(event streamingccl.Eve
...
-		return errors.New("kv event expected to have kv")
-	}
-	for _, kv := range kvs {
-		lrw.buffer.addKV(kv)
-	}
-	return nil
-}
-
-func (lrw *logicalReplicationWriterProcessor) bufferCheckpoint(event streamingccl.Event) error {
+func (lrw *logicalReplicationWriterProcessor) checkpoint(event streamingccl.Event) error {
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:395]: Should this take a ctx and select on ctx.Done() during the send? Or perhaps on the stopCh?

**Review by stevendanna — COMMENTED** [review]
I guess the fundamental question here is whether we ever want to batch more rows on the destination than we do on the source.

Replicator currently does fairly large insert batches (using more sophisticated SQL than we have here). That said, I think if we want to do similar batching, the buffer we use should probably be more sophisticated than this, so I'm still +1 on this.

The WithDiff PR is going to conflict with this pretty badly, but I don't mind being on the conflict side as this seems better to go in first if we are gonna do it.

**Comment by dt** [comment]
Why don’t you just merges with diff first; my merge strategy will just be `—mine` so shouldn’t be too bad

File: pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go (hunk 1 above)
Comment (dt) [reply pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:520]: we’re grabbing here before we sort, which I think is the same as it was before?

Comment (stevendanna) [reply pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:520]: Ohhhhhh, I completely forgot we are sorting by ts.

Comment (dt) [reply pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:520]: We’re not sorting by ts per se. we’re getting my the first key from the unsorted buffer before we sort it, which is thus still in rangefeed emit order.

**Comment by stevendanna** [comment]
> Why don’t you just merges with diff first; my merge strategy will just be —mine so shouldn’t be too bad

done.

File: pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go (hunk 2 above)
Comment (dt) [reply pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go:291]: done

**Review by msbutler — APPROVED** [review]

**Comment by dt** [comment]
TFTR!

bors r+

**Pull request merged**